src/pybind/mgr/mirroring/fs/dir_map/update.py

```diff
@@ -0,0 +1,151 @@
+import errno
+import pickle
+import logging
+
+import rados
+
+from ..utils import MIRROR_OBJECT_NAME, DIRECTORY_MAP_PREFIX, \
+    INSTANCE_ID_PREFIX, MIRROR_OBJECT_PREFIX
+
+log = logging.getLogger(__name__)
+
+MAX_UPDATE = 256
+
+class UpdateDirMapRequest(object):
+    def __init__(self, ioctx, update_mapping, removals, on_finish_callback):
+        self.ioctx = ioctx
+        self.update_mapping = update_mapping
+        self.removals = removals
+        self.on_finish_callback = on_finish_callback
+
+    @staticmethod
+    def omap_key(dir_path):
+        return f'{DIRECTORY_MAP_PREFIX}{dir_path}'
+
+    def send(self):
+        log.info('updating image map')
+        self.send_update()
+
+    def send_update(self):
+        log.debug(f'pending updates: {len(self.update_mapping)}+{len(self.removals)}')
+        try:
+            with rados.WriteOpCtx() as write_op:
+                keys = []
+                vals = []
+                dir_keys = list(self.update_mapping.keys())[0:MAX_UPDATE]
+                # gather updates
+                for dir_path in dir_keys:
+                    mapping = self.update_mapping.pop(dir_path)
+                    keys.append(UpdateDirMapRequest.omap_key(dir_path))
+                    vals.append(pickle.dumps(mapping))
+                    self.ioctx.set_omap(write_op, tuple(keys), tuple(vals))
+                # gather deletes
+                slicept = MAX_UPDATE - len(dir_keys)
+                removals = [UpdateDirMapRequest.omap_key(dir_path) for dir_path in self.removals[0:slicept]]
+                self.removals = self.removals[slicept:]
+                self.ioctx.remove_omap_keys(write_op, tuple(removals))
+                log.debug(f'applying {len(keys)} updates, {len(removals)} deletes')
+                self.ioctx.operate_aio_write_op(write_op, MIRROR_OBJECT_NAME, oncomplete=self.handle_update)
+        except rados.Error as e:
+            log.error(f'UpdateDirMapRequest.send_update exception: {e}')
+            self.finish(-e.args[0])
+
+    def handle_update(self, completion):
+        r = completion.get_return_value()
+        log.debug(f'handle_update: r={r}')
+        if not r == 0:
+            self.finish(r)
+        elif self.update_mapping or self.removals:
+            self.send_update()
+        else:
+            self.finish(0)
+
+    def finish(self, r):
+        log.info(f'finish: r={r}')
+        self.on_finish_callback(r)
+
+class UpdateInstanceRequest(object):
```

Comment:
```suggestion
class UpdateInstanceRequest:
```

in python3, all classes are new-style classes.

Reply:
ack